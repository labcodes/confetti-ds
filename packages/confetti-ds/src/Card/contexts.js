import { createContext } from "react";

export const CardContext = createContext();
