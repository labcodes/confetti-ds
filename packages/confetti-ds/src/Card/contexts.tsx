import { createContext } from "react";

export const CardContext = createContext<{
  color: string;
  skin?: string;
  cardType?: string;
}>(null);
