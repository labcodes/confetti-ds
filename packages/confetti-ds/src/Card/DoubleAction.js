import React, { useContext, useEffect } from "react";
import PropTypes from "prop-types";

import { Button, OutlineButton, TextButton } from "../Button";
import { CardContext } from "./contexts";

/**
 *
 * @param actionsProps sets button's attributes: label, action to be executed when the button is clicked, whether it is disabled/ariaDisabled or not, and the icon.
 * @param size sets buttons' height. Small = 32px, Normal = 40px, Large = 48px.
 * @param isText sets buttons to be Text Buttons.
 * @param isHorizontal sets whether buttons will pile on top of each other or be on the same line.
 * @returns {JSX.Element}
 * @constructor
 */

export default function DoubleAction({
  actionsProps,
  size,
  isText,
  isHorizontal,
}) {
  const cardContext = useContext(CardContext);
  const getButtonSkinFromCardContext = (context) => {
    const { color, skin, cardType } = context;
    if (cardType === "filled" && skin === "vivid" && color !== "white") {
      return "light";
    }

    if (
      cardType !== "outline" &&
      skin === "pale" &&
      (color === "purple" || color === "mineral")
    ) {
      return "dark";
    }
    return null;
  };

  const checkActionsPropsLength = (actionsPropsList) => {
    if (actionsPropsList.length !== 2) {
      throw Error(
        "DoubleAction: You need to pass exactly two objects inside the `actionsProps` prop."
      );
    }
  };

  useEffect(() => {
    checkActionsPropsLength(actionsProps);
  }, [actionsProps]);

  return (
    <section
      className={`lab-card-double-action lab-card-double-action--${size}
            ${isText ? " lab-card-double-action--text" : ""}
            ${isHorizontal ? " lab-card-double-action--horizontal" : ""}
          `}
    >
      {actionsProps.map(
        ({ text, onClick, icon, disabled, ariaDisabled }, index) => {
          if (isText) {
            return (
              <TextButton
                key={text}
                text={text}
                onClick={onClick}
                size={size}
                //  todo: reescrever para getContext
                skin={getButtonSkinFromCardContext(cardContext)}
                {...(icon ? { icon } : undefined)}
                {...(disabled ? { disabled } : undefined)}
                {...(ariaDisabled ? { ariaDisabled } : undefined)}
              />
            );
          }
          if (index === 0) {
            return (
              <Button
                key={text}
                text={text}
                onClick={onClick}
                size={size}
                skin={getButtonSkinFromCardContext(cardContext)}
                {...(icon ? { icon } : undefined)}
                {...(disabled ? { disabled } : undefined)}
                {...(ariaDisabled ? { ariaDisabled } : undefined)}
              />
            );
          }
          return (
            <OutlineButton
              key={text}
              text={text}
              onClick={onClick}
              size={size}
              skin={getButtonSkinFromCardContext(cardContext)}
              {...(icon ? { icon } : undefined)}
              {...(disabled ? { disabled } : undefined)}
              {...(ariaDisabled ? { ariaDisabled } : undefined)}
            />
          );
        }
      )}
    </section>
  );
}

DoubleAction.propTypes = {
  /** Sets button's attributes: label, action to be executed when the button is clicked, whether it is disabled/ariaDisabled or not, and the icon. */
  actionsProps: PropTypes.arrayOf(
    PropTypes.exact({
      text: PropTypes.string.isRequired,
      onClick: PropTypes.func.isRequired,
      ariaDisabled: PropTypes.bool,
      disabled: PropTypes.bool,
      icon: PropTypes.string,
    })
  ).isRequired,
  /** Sets buttons' height. Small = 32px, Normal = 40px, Large = 48px. */
  size: PropTypes.oneOf(["normal", "small", "large"]),
  /** Sets whether buttons will pile on top of each other or be on the same line. */
  isHorizontal: PropTypes.bool,
  /** Sets buttons to be Text Buttons. */
  isText: PropTypes.bool,
};

DoubleAction.defaultProps = {
  size: "normal",
  isHorizontal: false,
  isText: false,
};
