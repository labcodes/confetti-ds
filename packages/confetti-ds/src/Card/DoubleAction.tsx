import React, { SyntheticEvent, useContext, useEffect } from "react";

import { Button, OutlineButton, TextButton } from "../Button";
import { CardContext } from "./contexts";
import { IconTypes } from "../constants";
export interface DoubleActionProps {
  /** Sets button's attributes: label, action to be executed when the button is clicked, whether it is disabled/ariaDisabled or not, and the icon. */
  actionsProps: {
    label: string;
    onClick: (event?: SyntheticEvent) => any;
    ariaDisabled?: boolean;
    disabled?: boolean;
    icon?: IconTypes;
    hasIcon?: boolean;
  }[];
  /** Sets buttons' height. Small = 32px, Normal = 40px, Large = 48px. */
  size?: "normal" | "small" | "large";
  /** Sets whether buttons will pile on top of each other or be on the same line. */
  isHorizontal?: boolean;
  /** Sets buttons to be Text Buttons. */
  isText?: boolean;
}

export default function DoubleAction({
  actionsProps,
  size = "normal" as const,
  isHorizontal = false,
  isText = false,
}: DoubleActionProps) {
  const cardContext = useContext(CardContext);
  const getButtonSkinFromCardContext = (context) => {
    const { color, theme, cardType } = context;
    if (cardType === "filled" && theme === "vivid" && color !== "white") {
      return "light";
    }

    if (
      cardType !== "outline" &&
      theme === "pale" &&
      (color === "purple" || color === "mineral")
    ) {
      return "dark";
    }
    return null;
  };

  const checkActionsPropsLength = (actionsPropsList) => {
    if (actionsPropsList.length !== 2) {
      throw Error(
        "DoubleAction: You need to pass exactly two objects inside the `actionsProps` prop."
      );
    }
  };

  useEffect(() => {
    checkActionsPropsLength(actionsProps);
  }, [actionsProps]);

  return (
    <section
      className={`lab-card-double-action lab-card-double-action--${size}
            ${isText ? " lab-card-double-action--text" : ""}
            ${isHorizontal ? " lab-card-double-action--horizontal" : ""}
          `}
    >
      {actionsProps.map(
        ({ label, onClick, hasIcon, icon, disabled, ariaDisabled }, index) => {
          if (isText) {
            return (
              <TextButton
                key={label}
                label={label}
                onClick={onClick}
                size={size}
                theme={getButtonSkinFromCardContext(cardContext)}
                hasIcon
                {...(icon ? { icon } : undefined)}
                {...(disabled ? { disabled } : undefined)}
                {...(ariaDisabled ? { ariaDisabled } : undefined)}
              />
            );
          }
          if (index === 0) {
            return (
              <Button
                key={label}
                label={label}
                onClick={onClick}
                size={size}
                theme={getButtonSkinFromCardContext(cardContext)}
                hasIcon
                {...(icon ? { icon } : undefined)}
                {...(disabled ? { disabled } : undefined)}
                {...(ariaDisabled ? { ariaDisabled } : undefined)}
              />
            );
          }
          return (
            <OutlineButton
              key={label}
              label={label}
              onClick={onClick}
              size={size}
              theme={getButtonSkinFromCardContext(cardContext)}
              hasIcon
              {...(icon ? { icon } : undefined)}
              {...(disabled ? { disabled } : undefined)}
              {...(ariaDisabled ? { ariaDisabled } : undefined)}
            />
          );
        }
      )}
    </section>
  );
}
