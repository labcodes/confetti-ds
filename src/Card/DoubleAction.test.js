import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import DoubleAction from "./DoubleAction";
import { CardContext } from "./contexts";

describe("DoubleAction", () => {
  const originalError = console.error;
  afterEach(() => {
    console.error = originalError;
  });

  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(
        <CardContext.Provider value={{ color: "teal" }}>
          <DoubleAction
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("calls props.onClick when clicked", async () => {
    const mockOnClick1 = jest.fn();
    const mockOnClick2 = jest.fn();
    const mountedLinkAction = mount(
      <CardContext.Provider value={{ color: "teal" }}>
        <DoubleAction
          actionsProps={[
            {
              text: "Action 1",
              onClick: mockOnClick1,
            },
            {
              text: "Action 2",
              onClick: mockOnClick2,
            },
          ]}
        />
      </CardContext.Provider>
    );

    expect(mockOnClick1.mock.calls.length).toEqual(0);
    mountedLinkAction.find("button").first().simulate("click");
    expect(mockOnClick1.mock.calls.length).toEqual(1);

    expect(mockOnClick2.mock.calls.length).toEqual(0);
    mountedLinkAction.find("button").last().simulate("click");
    expect(mockOnClick2.mock.calls.length).toEqual(1);
  });

  it("throws error if actionsProps.length !== 2", async () => {
    console.error = jest.fn(); // silences output
    expect(() =>
      mount(
        <CardContext.Provider
          value={{ color: "teal", skin: "vivid", cardType: "filled" }}
        >
          <DoubleAction
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
    ).toThrow(
      "DoubleAction: You need to pass exactly two objects inside the `actionsProps` prop."
    );

    expect(() =>
      mount(
        <CardContext.Provider
          value={{ color: "teal", skin: "vivid", cardType: "filled" }}
        >
          <DoubleAction
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
              {
                text: "Action 3",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
    ).toThrow(
      "DoubleAction: You need to pass exactly two objects inside the `actionsProps` prop."
    );
  });

  it("renders with text buttons", async () => {
    const renderedComponent = renderer
      .create(
        <CardContext.Provider value={{ color: "teal" }}>
          <DoubleAction
            isText
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with isHorizontal", async () => {
    const renderedComponent = renderer
      .create(
        <CardContext.Provider value={{ color: "teal" }}>
          <DoubleAction
            isHorizontal
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with size", async () => {
    const renderedComponent = renderer
      .create(
        <CardContext.Provider value={{ color: "teal" }}>
          <DoubleAction
            size="small"
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders buttons with light skin", async () => {
    const renderedComponent = renderer
      .create(
        <CardContext.Provider
          value={{ color: "teal", skin: "vivid", cardType: "filled" }}
        >
          <DoubleAction
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders buttons with dark skin", async () => {
    let renderedComponent = renderer
      .create(
        <CardContext.Provider
          value={{ color: "purple", skin: "pale", cardType: "filled" }}
        >
          <DoubleAction
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    renderedComponent = renderer
      .create(
        <CardContext.Provider
          value={{ color: "mineral", skin: "pale", cardType: "filled" }}
        >
          <DoubleAction
            actionsProps={[
              {
                text: "Action 1",
                onClick: () => {},
              },
              {
                text: "Action 2",
                onClick: () => {},
              },
            ]}
          />
        </CardContext.Provider>
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });
});
