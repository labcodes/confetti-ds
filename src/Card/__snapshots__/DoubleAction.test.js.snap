// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoubleAction renders buttons with dark skin 1`] = `
<section
  className="lab-card-double-action lab-card-double-action--normal
            
            
          "
>
  <button
    className="lab-btn lab-btn--default lab-btn--normal lab-btn--dark"
    onClick={[Function]}
    type="button"
  >
    
    Action 1
  </button>
  <button
    className="lab-btn lab-btn--outline lab-btn--normal lab-btn--dark"
    onClick={[Function]}
    type="button"
  >
    
    Action 2
  </button>
</section>
`;

exports[`DoubleAction renders buttons with dark skin 2`] = `
<section
  className="lab-card-double-action lab-card-double-action--normal
            
            
          "
>
  <button
    className="lab-btn lab-btn--default lab-btn--normal lab-btn--dark"
    onClick={[Function]}
    type="button"
  >
    
    Action 1
  </button>
  <button
    className="lab-btn lab-btn--outline lab-btn--normal lab-btn--dark"
    onClick={[Function]}
    type="button"
  >
    
    Action 2
  </button>
</section>
`;

exports[`DoubleAction renders buttons with light skin 1`] = `
<section
  className="lab-card-double-action lab-card-double-action--normal
            
            
          "
>
  <button
    className="lab-btn lab-btn--default lab-btn--normal lab-btn--light"
    onClick={[Function]}
    type="button"
  >
    
    Action 1
  </button>
  <button
    className="lab-btn lab-btn--outline lab-btn--normal lab-btn--light"
    onClick={[Function]}
    type="button"
  >
    
    Action 2
  </button>
</section>
`;

exports[`DoubleAction renders with base props 1`] = `
<section
  className="lab-card-double-action lab-card-double-action--normal
            
            
          "
>
  <button
    className="lab-btn lab-btn--default lab-btn--normal"
    onClick={[Function]}
    type="button"
  >
    
    Action 1
  </button>
  <button
    className="lab-btn lab-btn--outline lab-btn--normal"
    onClick={[Function]}
    type="button"
  >
    
    Action 2
  </button>
</section>
`;

exports[`DoubleAction renders with isHorizontal 1`] = `
<section
  className="lab-card-double-action lab-card-double-action--normal
            
             lab-card-double-action--horizontal
          "
>
  <button
    className="lab-btn lab-btn--default lab-btn--normal"
    onClick={[Function]}
    type="button"
  >
    
    Action 1
  </button>
  <button
    className="lab-btn lab-btn--outline lab-btn--normal"
    onClick={[Function]}
    type="button"
  >
    
    Action 2
  </button>
</section>
`;

exports[`DoubleAction renders with size 1`] = `
<section
  className="lab-card-double-action lab-card-double-action--small
            
            
          "
>
  <button
    className="lab-btn lab-btn--default lab-btn--small"
    onClick={[Function]}
    type="button"
  >
    
    Action 1
  </button>
  <button
    className="lab-btn lab-btn--outline lab-btn--small"
    onClick={[Function]}
    type="button"
  >
    
    Action 2
  </button>
</section>
`;

exports[`DoubleAction renders with text buttons 1`] = `
<section
  className="lab-card-double-action lab-card-double-action--normal
             lab-card-double-action--text
            
          "
>
  <button
    className="lab-btn lab-btn--text lab-btn--normal"
    onClick={[Function]}
    type="button"
  >
    
    Action 1
  </button>
  <button
    className="lab-btn lab-btn--text lab-btn--normal"
    onClick={[Function]}
    type="button"
  >
    
    Action 2
  </button>
</section>
`;
